File: VPFlowVueApp/src/lib/logging.ts
import { useSettings } from "@/composables/useSettings";

const MAX_LOG_LINES = 50;

function format(data: any[]) {
  return data
    .map((d) => (typeof d === "string" ? `"${d}"` : JSON.stringify(d)))
    .join(" ");
}

function trim(text: string) {
  const lines = text.split(/\r?\n/);
  if (lines.length <= MAX_LOG_LINES) {
    return text;
  }
  return lines.slice(-MAX_LOG_LINES).join("\r\n");
}

function commit(prefix: string, data: any[]) {
  const { logs } = useSettings();
  const entry = prefix + format(data);
  const newText = !logs.value ? entry : logs.value + "\r\n" + entry;
  logs.value = trim(newText);
}

export function log(...data: any[]) {
  console.log(...data);
  commit("> ", data);
}

export function warn(...data: any[]) {
  console.warn(...data);
  commit("[!] ", data);
}

export function error(...data: any[]) {
  console.error(...data);
  commit("[ERROR] ", data);
}

File: VPFlowWebMain/Logging.cs
using System;
using System.Diagnostics;
using System.Linq;
using System.Windows.Forms;

namespace VPFlowWebMain
{
    internal static class Logging
    {
        private static TextBox LogArea => MainForm.Instance.LogArea;

        private const int MAX_LOG_LINES = 50;

        private static void Commit(string prefix, string text)
        {
            // build new entry and append to existing text
            var entry = prefix + text;
            var current = string.IsNullOrEmpty(LogArea.Text) ? entry : (LogArea.Text + "\r\n" + entry);

            // trim to last MAX_LOG_LINES lines
            var lines = current.Split(new[] { "\r\n" }, StringSplitOptions.None);
            if (lines.Length > MAX_LOG_LINES)
            {
                current = string.Join("\r\n", lines.Skip(lines.Length - MAX_LOG_LINES));
            }

            LogArea.Text = current;

            // move caret to end and ensure visible
            LogArea.SelectionStart = LogArea.Text.Length;
            LogArea.ScrollToCaret();
        }

        public static void Log(string text)
        {
            Debug.WriteLine(text);
            Commit("> ", text);
        }

        public static void Warn(string text)
        {
            Debug.WriteLine(text);
            Commit("[!] ", text);
        }

        public static void Error(string text)
        {
            Debug.WriteLine(text);
            Commit("[ERROR] ", text);
        }
    }
}
